websockify/websocket.py

```diff
@@ -1,1109 +1,825 @@
-            self.msg("  - Backgrounding (daemon)")
-        if self.record:
-            self.msg("  - Recording to '%s.*'", self.record)
-
-    #
-    # WebSocketServer static methods
-    #
-
-    @staticmethod
-    def get_logger():
-        return logging.getLogger("%s.%s" % (
-            WebSocketServer.log_prefix,
-            WebSocketServer.__class__.__name__))
-
-    @staticmethod
-    def socket(host, port=None, connect=False, prefer_ipv6=False,
-               unix_socket=None, use_ssl=False, tcp_keepalive=True,
-               tcp_keepcnt=None, tcp_keepidle=None, tcp_keepintvl=None):
-        """ Resolve a host (and optional port) to an IPv4 or IPv6
-        address. Create a socket. Bind to it if listen is set,
-        otherwise connect to it. Return the socket.
+        Both WebSocketWantReadError and WebSocketWantWriteError can be
+        raised when calling recvmsg().
         """
-        flags = 0
-        if host == '':
-            host = None
-        if connect and not (port or unix_socket):
-            raise Exception("Connect mode requires a port")
-        if use_ssl and not ssl:
-            raise Exception("SSL socket requested but Python SSL module not loaded.");
-        if not connect and use_ssl:
-            raise Exception("SSL only supported in connect mode (for now)")
-        if not connect:
-            flags = flags | socket.AI_PASSIVE
-
-        if not unix_socket:
-            addrs = socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM,
-                    socket.IPPROTO_TCP, flags)
-            if not addrs:
-                raise Exception("Could not resolve host '%s'" % host)
-            addrs.sort(key=lambda x: x[0])
-            if prefer_ipv6:
-                addrs.reverse()
-            sock = socket.socket(addrs[0][0], addrs[0][1])
-
-            if  tcp_keepalive:
-                sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
-                if tcp_keepcnt:
-                    sock.setsockopt(socket.SOL_TCP, socket.TCP_KEEPCNT,
-                                    tcp_keepcnt)
-                if tcp_keepidle:
-                    sock.setsockopt(socket.SOL_TCP, socket.TCP_KEEPIDLE,
-                                    tcp_keepidle)
-                if tcp_keepintvl:
-                    sock.setsockopt(socket.SOL_TCP, socket.TCP_KEEPINTVL,
-                                    tcp_keepintvl)
-
-            if connect:
-                sock.connect(addrs[0][4])
-                if use_ssl:
-                    sock = ssl.wrap_socket(sock)
-            else:
-                sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-                sock.bind(addrs[0][4])
-                sock.listen(100)
-        else:
-            sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
-            sock.connect(unix_socket)
+        # May have been called to flush out a close
+        if self._received_close:
+            self._flush()
+            return ''.encode("ascii")
```

Comment:
An empty buffer is usually how `recv()` indicates a nicely closed socket. Anything else seems more like an abnormal condition. `None` can also be the result of a stray `return`.